client/task_dispatcher.go

```diff
@@ -0,0 +1,212 @@
+package client
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/status"
+	"github.com/google/uuid"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+
+	"github.com/hanfei1991/microcosm/pb"
+	derrors "github.com/hanfei1991/microcosm/pkg/errors"
+)
+
+const preDispatchTaskRetryInterval = 1 * time.Second
+
+// TaskDispatcher implements the logic to invoke two-phase task-dispatching.
+// A separate struct is used to decouple the complexity of the two-phase
+// protocol from the implementation of ExecutorClient.
+// TODO think about whether we should refactor the ExecutorClient's interface.
+type TaskDispatcher struct {
+	client ExecutorClient
+
+	timeout       time.Duration
+	retryInterval time.Duration
+}
+
+// NewTaskDispatcher returns a new TaskDispatcher.
+// timeout limits the total duration of a call to DispatchTask.
+func NewTaskDispatcher(client ExecutorClient, timeout time.Duration) *TaskDispatcher {
+	return &TaskDispatcher{
+		client:        client,
+		timeout:       timeout,
+		retryInterval: preDispatchTaskRetryInterval,
+	}
+}
+
+// DispatchTaskArgs contains the required parameters for creating a worker.
+type DispatchTaskArgs struct {
+	WorkerID     string
+	MasterID     string
+	WorkerType   int64
+	WorkerConfig []byte
+}
+
+// DispatchTask performs the two-phase task dispatching by
+// doing the relevant gRPC calls.
+//
+// - startWorkerTimer() is called before a worker is
+//   launched on the executor.
+// - abortWorker() is called if startWorkerTimer() has been called
+//   but the server has aborted the request.
+//
+// For a sketch of the logic, please refer to lib/doc/create_worker.puml.
+// TODO The UML diagrams are not finalized.
+// TODO Find a way to automatically render the puml files on GitHub.
+func (d *TaskDispatcher) DispatchTask(
+	ctx context.Context,
+	args *DispatchTaskArgs,
+	startWorkerTimer func(),
+	abortWorker func(),
+) error {
+	ctx, cancel := context.WithTimeout(ctx, d.timeout)
+	defer cancel()
+
+	requestID, err := d.preDispatchTaskWithRetry(ctx, args)
+	if err != nil {
+		return derrors.ErrExecutorPreDispatchFailed.Wrap(err)
+	}
+
+	// The timer must be started before invoking ConfirmDispatchTask.
+	startWorkerTimer()
```

Comment:
Is this timer used for checking worker timeout after `confirmDispatchTask` returns success or returns undetermined error. Why it is must be started before invoking `confirmDispatchTask`

Reply:
- The timer is a virtual concept. `startWorkerTimer` actually corresponds to creating a `workerEntry` with an `ExpireAt` counting from now.
- It must be called before invoking `confirmDispatchTask` is because the Runtime will start the worker once it receives `ConfirmDispatchTask`. So we are expect heartbeats soon after calling `ConfirmDispatchTask`. To make sure no heartbeat is missed, or other unexpected problems, we should call `startWorkerTimer` before invoking `ConfirmDispatchTask`.